framework/include/userobjects/TimedElementSubdomainModifier.h

```diff
@@ -0,0 +1,69 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ElementSubdomainModifier.h"
+
+/**
+ * Modifies element subdomains only at a given list of times
+ */
+class TimedElementSubdomainModifier : public ElementSubdomainModifier
+{
+public:
+  static InputParameters validParams();
+
+  TimedElementSubdomainModifier(const InputParameters & parameters);
+
+  virtual void initialize() override;
+
+protected:
+  virtual SubdomainID computeSubdomainID() override;
+
+  /**
+   * Requests a vector of all points in time from the inheriting class
+   * (these do not have to be sorted).
+   * @returns Unsorted vector of points in time.
+   */
+  virtual std::vector<Real> onGetTimes();
+
+  /**
+   * Determines the new subdomain-id for the current element.
+   * @returns New subdomain id to move the element to.
+   */
+  virtual SubdomainID onComputeSubdomainID(Real t_from_exclusive, Real t_to_inclusive);
```

Comment:
The name of this method doesn't agree very well with what it actually does. If I were overriding this method, I don't quite understand what to return, and what the two arguments represent.

Reply:
The most confusing part to me is the time arguments. Which leads me to my question of why we need all this manual tracking?
```c++
  // did we advance to the next step?                                                                                     
  if (_current_step != unsigned(_t_step))
  {
    _current_timespan_start = _current_timespan_end;
    _current_timespan_end = _t;
    _current_step = _t_step;
  }
```
If you reach a time point indicated by the user, isn't that the signal to do your subdomain modification? I feel like if this method needs to exist, it should only take one argument. I don't know why we need both `_current_timestep_start` and `_current_timestep_end` or why `_t_step` itself is insufficient such that we need `_current_step`? I was hoping to help you address @hugary1995's suggested changes and most of them are straightforward, but this is a stickier point that I need some clarity on

Reply:
I would need to double check but I think the manual accounting is to handle the case where several subdomain changes step happen (possibly doing block id 1-> block id 2 then 2->3) inside a single time step

Reply:
As @GiudGiud wrote, this is to handle multiple subdomain changes for a single time step. 

I have also implemented this manual accounting to keep the block assignments consistent, even if the user skips / misplaces some time steps in the [Executioner] block.

Of course, if the user chooses the timesteps “perfectly”, there would be no need for this manual accounting. On the other hand, with the manual accounting the numerical effort is minimal and Moose models with a TimedElementSubdomainModifier are much more robust against “clumsy” specifications in the time steps.

Reply:
The data members still don't make sense to me. To me it sounds like you should have some map from time step to a sorted vector of pairs, where the vector of pairs represents the sequence of subdomain modifications with the pairs being from-to

Reply:
Dear @lindsayad, 

As requested [here](https://github.com/idaholab/moose/discussions/26952#discussioncomment-8694525) by @hugary1995, I introduced TimedElementSubdomainModifier to be inherited from as a intermediate class. The functionality itself including the mapping is to be found in the TimedSubdomainModifier. 

So for me this "mapping" is stored in the 3 variables, in [TimedSubdomainModifier.h](https://github.com/idaholab/moose/pull/27092/files?short_path=90ce7a0#diff-7f4ac19421b805698585d4de8a6dd0e9ee3191be38fa346bd02dc5e0ae36243c) lines 39 to 44:
std::vector<Real> _times;
std::vector<SubdomainID> _blocks_from;
std::vector<SubdomainID> _blocks_to;

Reply:
Ok I finally understood that the timespans were meant to handle bracketing between `_t` and `_t_old` (right?). I've added a `_t_old` data member to the `TransientInterface` which should hopefully make this much more straightforward to handle

Reply:
And consequently I don't think we need a timed base class implementation of `computeSubdomainID`. I think everything can be handled in a derived class implementation of `computeSubdomainID` which removes the need for this `onComputeSubdomainID` method

Reply:
Dear @lindsayad, Thank you for your support. With the `_t_old` data member to the `TransientInterface` the code gets a lot clearer. Thanks to your code I'll resove this conversation.